openlibrary/core/ratings.py

```diff
@@ -0,0 +1,94 @@
+from . import db
+from openlibrary.core.bookshelves import Bookshelves
+
+class Ratings(object):
+
+    VALID_STAR_RATINGS = range(11)  # inclusive: [0 - 10] (0-5 star, w/ half stars)
+
+    @classmethod
+    def total_num_books_rated(cls, since=None, distinct=False):
+        oldb = db.get_db()
+        query = "SELECT count(%s work_id) from ratings" % ('DISTINCT' if distinct else '')
+        if since:
+            query += " WHERE created >= $since"
+        results = oldb.query(query, vars={'since': since})
+        return results[0] if results else None
+
+    @classmethod
+    def total_num_unique_raters(cls, since=None):
+        oldb = db.get_db()
+        query = "select count(DISTINCT username) from ratings"
+        if since:
+            query += " WHERE created >= $since"
+        results = oldb.query(query, vars={'since': since})
+        return results[0] if results else None
+
+    @classmethod
+    def most_rated_books(cls, limit=10, since=False):
+        oldb = db.get_db()
+        query = ('select work_id, count(*) as cnt from ratings group '
+                 'by work_id order by cnt desc limit $limit')
+        if since:
+            query += " WHERE created >= $since"
+        return list(oldb.query(query, vars={'limit': limit, 'since': since}))
+
+    @classmethod
+    def get_users_ratings(cls, username):
+        oldb = db.get_db()
+        query = 'select * from ratings where username=$username'
+        return list(oldb.query(query, vars={'username': username}))
+
+    @classmethod
+    def get_all_works_ratings(cls, work_id):
+        oldb = db.get_db()
+        query = 'select * from ratings where work_id=$work_id'
+        return list(oldb.query(query, vars={'work_id': work_id}))
+
+    @classmethod
+    def get_users_rating_for_work(cls, username, work_id):
+        oldb = db.get_db()
+        data = {
+            'username': username,
+            'work_id': int(work_id)
+        }
+        query = 'SELECT * from ratings where username=$username AND work_id=$work_id'
+        results = list(oldb.query(query, vars=data))
+        rating = results[0].rating if results else None
+        return rating
+
+    @classmethod
+    def remove(cls, username, work_id):
+        oldb = db.get_db()
+        where = {
+            'username': username,
+            'work_id': int(work_id)
+        }
+        try:
+            return oldb.delete('ratings', where=(
+                'work_id=$work_id AND username=$username'), vars=where)
```

Comment:
See: http://webpy.org/cookbook/query

To prevent SQL injection attacks, db.query also accepts the “vars” syntax as described in db.select:
```
results = db.query("SELECT * FROM users WHERE id=$id", vars={'id':10})
```
This will escape user input, if you’re trusting them for the “id” variable.